static/app/components/dropdownMenu.tsx

```diff
@@ -105,43 +105,51 @@ type Props = DefaultProps & {
   children: (renderProps: RenderProps) => React.ReactNode;
 };
 
-type State = {
-  isOpen: boolean;
-};
-
-class DropdownMenu extends React.Component<Props, State> {
-  static defaultProps: DefaultProps = {
-    keepMenuOpen: false,
-    closeOnEscape: true,
-  };
-
-  state: State = {
-    isOpen: false,
-  };
-
-  componentWillUnmount() {
-    document.removeEventListener('click', this.checkClickOutside, true);
-  }
-
-  dropdownMenu: Element | null = null;
-  dropdownActor: Element | null = null;
-
-  mouseLeaveId: number | null = null;
-  mouseEnterId: number | null = null;
+function DropdownMenu({
+  keepMenuOpen = false,
+  closeOnEscape = true,
+  onOpen,
+  onClose,
+  onClickOutside,
+  shouldIgnoreClickOutside,
+  isOpen: isOpenProp,
+  alwaysRenderMenu,
+  isNestedDropdown,
+  children,
+}: Props) {
+  const dropdownMenu = useRef<Element | null>(null);
+  const dropdownActor = useRef<Element | null>(null);
+
+  let mouseLeaveId: number | null = null;
+  let mouseEnterId: number | null = null;
 
-  // Gets open state from props or local state when appropriate
-  isOpen = () => {
-    const {isOpen} = this.props;
-    const isControlled = typeof isOpen !== 'undefined';
-    return (isControlled && isOpen) || this.state.isOpen;
-  };
-
-  // Checks if click happens inside of dropdown menu (or its button)
-  // Closes dropdownmenu if it is "outside"
-  checkClickOutside = async (e: MouseEvent) => {
-    const {onClickOutside, shouldIgnoreClickOutside} = this.props;
+  /**
+   * Internal isOpen state, separate from isOpenProp. This is used when
+   * component is not controlled, i.e. there is no isOpenProp passed
+   */
+  const [isOpenInternalState, setIsOpenInternalState] = useState(false);
+  /**
+   * Combined isOpen state, set to isOpenProp if component
+   * is controlled, or to isOpenState otherwise
+   */
+  const [isOpen, setIsOpen] = useState(false);
+  /**
+   * Parallel ref to isOpen (should always be equal to the isOpen state),
+   * to access isOpen state from event listeners, like checkClickOutside.
+   */
+  const isOpenRef = useRef(false);
+  useEffect(() => {
+    const newIsOpen = isOpenProp ?? isOpenInternalState;
+    setIsOpen(newIsOpen);
+    isOpenRef.current = newIsOpen;
+  }, [isOpenProp, isOpenInternalState]);
 
-    if (!this.dropdownMenu || !this.isOpen()) {
+  /**
+   * Checks if click happens inside of dropdown menu (or its button)
+   * Closes dropdownmenu if it is "outside"
+   */
+  const checkClickOutside = async (e: MouseEvent) => {
```

Comment:
I think you may need to `useCallback` on this to ensure that we keep the same reference to the function across renders, so that we can properly remove the listener when unmounting.

cc @JonasBa could you double check me here?

Reply:
yes indeed, if somehow the component rerenders without the removeEventListener being called, we'll never cleanup the callback. I would tipically do both of these in a single useEffect, that keeps the code tied together and easy to maintain

```typescript
const checkClickOutside = useCallback(() => ...,[])
useEffect(() => {
    document.addEventListener("click", checkClickOutside, true)
    return () => document.removeEventListener('click', checkClickOutside, true);
  }, [checkClickOutside]);
```